From 34d0bf40344509910e7dc2391a5e2d947d9b9487 Mon Sep 17 00:00:00 2001
From: Ofer Heifetz <oferh@marvell.com>
Date: Wed, 11 Nov 2015 16:22:36 +0200
Subject: [PATCH 0072/1345] thermal: armada: change temperature band gap
 circuit curve

commit  4ff2d8a7d37fa083515e9d8ae6028fd310e34086 from
https://github.com/MarvellEmbeddedProcessors/linux-marvell.git

a38x default used curve introduces temperature read to be
incorrectly high.

Signed-off-by: Ofer Heifetz <oferh@marvell.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/24790
Tested-by: Star_Automation <star@marvell.com>
Reviewed-by: Lior Amsalem <alior@marvell.com>

Conflicts:
	arch/arm/boot/dts/armada-38x.dtsi

Change-Id: I9461d5b4ab913527674ffa92effa351201a89f27
Signed-off-by: Grzegorz Jaszczyk <jaz@semihalf.com>
Reviewed-on: http://vgitil04.il.marvell.com:8080/27579
Reviewed-by: Ofer Heifetz <oferh@marvell.com>
Tested-by: Ofer Heifetz <oferh@marvell.com>
Signed-off-by: Meng Li <Meng.Li@windriver.com>
---
 arch/arm/boot/dts/armada-38x.dtsi |    2 +-
 drivers/thermal/armada_thermal.c  |   13 +++++++++++--
 2 files changed, 12 insertions(+), 3 deletions(-)

diff --git a/arch/arm/boot/dts/armada-38x.dtsi b/arch/arm/boot/dts/armada-38x.dtsi
index 093bad0..40f4efe 100644
--- a/arch/arm/boot/dts/armada-38x.dtsi
+++ b/arch/arm/boot/dts/armada-38x.dtsi
@@ -585,7 +585,7 @@
 
 			thermal: thermal@e8078 {
 				compatible = "marvell,armada380-thermal";
-				reg = <0xe4078 0x4>, <0xe4074 0x4>;
+				reg = <0xe4078 0x4>, <0xe4070 0x8>;
 				status = "okay";
 			};
 
diff --git a/drivers/thermal/armada_thermal.c b/drivers/thermal/armada_thermal.c
index 4e3f2d4..1d5c57b 100644
--- a/drivers/thermal/armada_thermal.c
+++ b/drivers/thermal/armada_thermal.c
@@ -40,6 +40,8 @@
 #define A375_READOUT_INVERT		BIT(15)
 #define A375_HW_RESETn			BIT(8)
 #define A380_HW_RESET			BIT(8)
+#define A380_CONTROL_MSB_OFFSET		4
+#define A380_TSEN_TC_TRIM_MASK		0x7
 
 #define AP806_START	BIT(0)
 #define AP806_RESET	BIT(1)
@@ -148,14 +150,21 @@ static void armada375_init_sensor(struct platform_device *pdev,
 static void armada380_init_sensor(struct platform_device *pdev,
 				  struct armada_thermal_priv *priv)
 {
-	unsigned long reg = readl_relaxed(priv->control);
+	unsigned long reg = readl_relaxed(priv->control +
+					  A380_CONTROL_MSB_OFFSET);
 
 	/* Reset hardware once */
 	if (!(reg & A380_HW_RESET)) {
 		reg |= A380_HW_RESET;
-		writel(reg, priv->control);
+		writel(reg, priv->control + A380_CONTROL_MSB_OFFSET);
 		mdelay(10);
 	}
+
+	/* set Tsen Tc Trim value */
+	reg = readl_relaxed(priv->control);
+	reg &= ~A380_TSEN_TC_TRIM_MASK;
+	reg |= 0x3;
+	writel(reg, priv->control);
 }
 
 static void armada_ap806_init_sensor(struct platform_device *pdev,
-- 
1.7.9.5

